File: iMods/Classes/Models/Item.m
//
//  Item.m
//  iMods
//
//  Created by Ryan Feng on 7/17/14.
//  Copyright (c) 2014 Ryan Feng. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <Mantle/NSValueTransformer+MTLPredefinedTransformerAdditions.h>
#import <Mantle/MTLValueTransformer.h>
#include <Mantle/MTLModel.h>
#include <Mantle/MTLJSONAdapter.h>

@interface IMOItem : MTLModel <MTLJSONSerializing>

/* JSON data fields */

@property (nonatomic, copy, readonly) NSNumber * item_id;
@property (nonatomic, copy, readonly) NSNumber * category_id;
@property (nonatomic, copy, readonly) NSString * author_id;
@property (nonatomic, copy, readonly) NSString * pkg_name;
@property (nonatomic, copy, readonly) NSString * pkg_version;
@property (nonatomic, copy, readonly) NSString * pkg_assets_path;
@property (nonatomic, copy, readonly) NSString * pkg_signature;
@property (nonatomic, copy, readonly) NSString * pkg_dependencies;
@property (nonatomic, copy, readonly) NSString * display_name;
@property (nonatomic, assign, readonly) float price;
@property (nonatomic, copy, readonly) NSString * summary;
@property (nonatomic, copy, readonly) NSString * description;
@property (nonatomic, copy, readonly) NSDate * add_date;
@property (nonatomic, copy, readonly) NSDate * last_update_date;

/* Non-JSON data fields */

@end

@implementation IMOItem

+ (NSDictionary *) JSONKeyPathsByPropertyKey {
    return @{
             @"iid": @"item_id",
             @"category_id": @"category_id",
             @"author_id": @"author_id",
             @"pkg_name": @"pkg_name",
             @"pkg_version": @"pkg_version",
             @"pkg_assets_path": @"pkg_assets_path",
             @"pkg_signature": @"pkg_signature",
             @"pkg_dependencies": @"pkg_dependencies",
             @"display_name": @"display_name",
             @"price": @"price",
             @"summary": @"summary",
             @"description": @"description",
             @"add_date": @"add_date",
             @"last_update_date": @"last_update_date"
             };
}

+ (NSDateFormatter*) dateFormatter{
    NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
    dateFormatter.locale = [[NSLocale alloc] initWithLocaleIdentifier:@"en_US_POSIX"];
    [dateFormatter setTimeZone:[NSTimeZone timeZoneWithAbbreviation:@"UTC"]];
    dateFormatter.dateFormat = @"yyyy-MM-dd'T'HH:mm:ss'Z'";
    return dateFormatter;
}

+ (NSValueTransformer*) addDateJSONTransformer {
    return [MTLValueTransformer reversibleTransformerWithForwardBlock:^(NSString* str) {
        return [self.dateFormatter dateFromString:str];
    }reverseBlock:^(NSDate* date){
        return [self.dateFormatter stringFromDate:date];
    }];
}

@end

File: iMods/Classes/Models/Item.swift
//
//  IMOItemModel.swift
//  iMods
//
//  Created by Brendon Roberto on 7/14/14.
//  Copyright (c) 2014 Brendon Roberto. All rights reserved.
//

import Foundation

class Item {
    var categoryId: Int
    var authorId: Int
    var packageName: String
    var displayName: String
    var packageVersion: String
    var packageSignature: PackageSignature?
    var packagePath: NSURL
    var packageAssetsPath: NSURL
    var packageDependencies: String[]?
    var price: Float
    var summary: String
    var description: String
    var addDate: NSDate
    var lastUpdateDate: NSDate

    init(categoryId: Int, authorId: Int, packageName: String, displayName: String,
        packageVersion: String, packageSignature: PackageSignature?,
        packagePath: NSURL, packageAssetsPath: NSURL,
        packageDependencies: String[]?, price: Float, summary: String,
        description: String, addDate: NSDate, lastUpdateDate: NSDate) {
            self.categoryId = categoryId
            self.authorId = authorId
            self.packageName = packageName
            self.displayName = displayName
            self.packageVersion = packageVersion
            self.packageSignature = packageSignature
            self.packagePath = packagePath
            self.packageAssetsPath = packageAssetsPath
            self.packageDependencies = packageDependencies
            self.price = price
            self.summary = summary
            self.description = description
            self.addDate = addDate
            self.lastUpdateDate = lastUpdateDate
    }
}